From 53d3f84369224716d241ee02f8461a40b8fe8a54 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Tue, 11 Mar 2025 15:00:08 +0100
Subject: [PATCH 1613/1695] drivers: remoteproc: reinitialize rproc fields in
 stm32_rproc_clean

rproc->boot_addr and rproc->tzen are set when we apply TLV configuration
on firmware image loading, and reset when we stop the remote processor.

If the load fails, we do not reinitialize the values as rproc_stop() is
never called. When we start a new firmware, we return an error because
the values are set twice.

Adding the variable reset in stm32_rproc_clean ensures that we return to
an initial state ready for a new load.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 93ee5c059549cb0e3f72bcb0aedcd75bdc5ed49b drivers: remoteproc: reinitialize rproc fields in stm32_rproc_clean]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: Ic1ecb6990aebadb9fca2b48858650bfc1a8e9859
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/443460
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 core/drivers/remoteproc/stm32_remoteproc.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/core/drivers/remoteproc/stm32_remoteproc.c b/core/drivers/remoteproc/stm32_remoteproc.c
index c679b836f..c99a47fe3 100644
--- a/core/drivers/remoteproc/stm32_remoteproc.c
+++ b/core/drivers/remoteproc/stm32_remoteproc.c
@@ -534,6 +534,9 @@ TEE_Result stm32_rproc_clean(uint32_t rproc_id)
 	if (stm32_rproc_release_mems_access(rproc) != TEE_SUCCESS)
 		panic();
 
+	rproc->boot_addr = 0;
+	rproc->tzen = false;
+
 	return res;
 }
 
-- 
2.47.2

